packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -860,10 +1231,119 @@ class RequireLtGadget : public GadgetT
     }
 };
 
-// (value * numerator) = product
-// product / denominator = quotient
-// product % denominator = remainder
-class MulDivGadget : public GadgetT
+// if (C) then require(A)
+class IfThenRequireGadget : public GadgetT
+{
+public:
+    NotGadget notC;
+    OrGadget res;
+
+    IfThenRequireGadget(
+        ProtoboardT& pb,
+        const VariableT& C,
+        const VariableT& A,
+        const std::string& prefix
+    ) :
+        GadgetT(pb, prefix),
+
+        notC(pb, C, FMT(prefix, ".notC")),
+        res(pb, {notC.result(), A}, FMT(prefix, ".res"))
+    {
+
+    }
+
+    void generate_r1cs_witness()
+    {
+        notC.generate_r1cs_witness();
+        res.generate_r1cs_witness();
+    }
+
+    void generate_r1cs_constraints()
+    {
+        notC.generate_r1cs_constraints();
+        res.generate_r1cs_constraints();
+        pb.add_r1cs_constraint(ConstraintT(res.result(), FieldT::one(), FieldT::one()), FMT(annotation_prefix, ".valid"));
+    }
+};
+
+// if (C) then require(A == B)
+class IfThenRequireEqualGadget : public GadgetT
+{
+public:
+    EqualGadget eq;
+    IfThenRequireGadget res;
+
+    IfThenRequireEqualGadget(
+        ProtoboardT& pb,
+        const VariableT& C,
+        const VariableT& A,
+        const VariableT& B,
+        const std::string& prefix
+    ) :
+        GadgetT(pb, prefix),
+
+        eq(pb, A, B, FMT(prefix, ".eq")),
+        res(pb, C, eq.result(), FMT(prefix, ".res"))
+    {
+
+    }
+
+    void generate_r1cs_witness()
+    {
+        eq.generate_r1cs_witness();
+        res.generate_r1cs_witness();
+    }
+
+    void generate_r1cs_constraints()
+    {
+        eq.generate_r1cs_constraints();
+        res.generate_r1cs_constraints();
+    }
+};
+
+// if (C) then require(A != B)
+class IfThenRequireNotEqualGadget : public GadgetT
+{
+public:
+    EqualGadget eq;
+    NotGadget notEq;
+    IfThenRequireGadget res;
+
+    IfThenRequireNotEqualGadget(
+        ProtoboardT& pb,
+        const VariableT& C,
+        const VariableT& A,
+        const VariableT& B,
+        const std::string& prefix
+    ) :
+        GadgetT(pb, prefix),
+
+        eq(pb, A, B, FMT(prefix, ".eq")),
+        notEq(pb, eq.result(), FMT(prefix, ".notEq")),
+        res(pb, C, notEq.result(), FMT(prefix, ".res"))
+    {
+
+    }
+
+    void generate_r1cs_witness()
+    {
+        eq.generate_r1cs_witness();
+        notEq.generate_r1cs_witness();
+        res.generate_r1cs_witness();
+    }
+
+    void generate_r1cs_constraints()
+    {
+        eq.generate_r1cs_constraints();
+        notEq.generate_r1cs_constraints();
+        res.generate_r1cs_constraints();
+    }
+};
+
+// (value * numerator) = product
+// product / denominator = quotient
+// product % denominator = remainder
+class MulDivGadget : public GadgetT
```

Comment:
It uses `UnsafeMulGadget` internally because if you break it out you may indeed need to make a safe variant because it may generate a bigger value than expected because it may get misused. By making this an atomic operation with the division we can be sure that isn't possible.

`product / denominator = quotient` and one for `product % denominator = remainder` need to do exactly the same thing, you need both the quotient and the remainder for the gadget to work, so seems like they will be the exact same gadget, just not exposing some of the values. The remainder is also never really used in the circuits, it's just exposed on the gadget because it needs to be there any way. So I don't think you can break the division up in different gadgets in a meaningful way.

This gadget is only used in `FeeCalculatorGadget` so kind of specifically written for that use case. I actually thought there were some other places as well which used the gadget the same way, but with the simplification of the matching logic those were removed.